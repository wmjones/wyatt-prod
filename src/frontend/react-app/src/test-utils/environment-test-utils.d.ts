/**
 * Type definitions for environment testing utilities
 */

import { SpyInstance } from 'jest';
import * as environmentModule from '../utils/environment';

/**
 * Mock the environment as development
 * @returns Jest spy instance to manage the mock
 */
export function mockDevelopmentEnvironment(): SpyInstance;

/**
 * Mock the environment as production
 * @returns Jest spy instance to manage the mock
 */
export function mockProductionEnvironment(): SpyInstance;

/**
 * Mock specific environment function with custom value
 * @param functionName The environment function to mock
 * @param value The return value for the function
 * @returns Jest spy instance to manage the mock
 */
export function mockEnvironmentFunction<T extends keyof typeof environmentModule>(
  functionName: T,
  value: boolean | string
): SpyInstance;

/**
 * Reset all environment mocks
 */
export function resetEnvironmentMocks(): void;
