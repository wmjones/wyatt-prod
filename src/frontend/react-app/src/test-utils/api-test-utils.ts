import { server } from '../mocks/server';
import { rest } from 'msw';

/**
 * Utility function to override MSW handlers for a specific test
 * @param path API path to override
 * @param method HTTP method (get, post, etc.)
 * @param responseData Data to return in the response
 * @param status HTTP status code
 */
export const mockApiResponse = (
  path: string,
  method: 'get' | 'post' | 'put' | 'delete' | 'patch',
  responseData: any,
  status = 200
) => {
  // Create a handler for the specified method
  const handler = rest[method](
    path,
    (req, res, ctx) => {
      return res(
        ctx.status(status),
        ctx.json(responseData)
      );
    }
  );

  // Use MSW to override the default handler
  server.use(handler);
};

/**
 * Utility function to simulate API error responses
 * @param path API path to override
 * @param method HTTP method
 * @param status HTTP error status code
 * @param errorData Optional error data
 */
export const mockApiError = (
  path: string,
  method: 'get' | 'post' | 'put' | 'delete' | 'patch',
  status = 500,
  errorData = { message: 'An error occurred' }
) => {
  // Create an error handler
  const handler = rest[method](
    path,
    (req, res, ctx) => {
      return res(
        ctx.status(status),
        ctx.json(errorData)
      );
    }
  );

  // Use MSW to override the default handler
  server.use(handler);
};

/**
 * Utility function to simulate network failures
 * @param path API path to override
 * @param method HTTP method
 */
export const mockNetworkError = (
  path: string,
  method: 'get' | 'post' | 'put' | 'delete' | 'patch'
) => {
  // Create a handler that simulates a network error
  const handler = rest[method](
    path,
    (req, res, ctx) => {
      return res(
        ctx.status(503),
        ctx.json({ message: 'Network Error' })
      );
    }
  );

  // Use MSW to override the default handler
  server.use(handler);
};

/**
 * Reset all handlers to the default handlers
 */
export const resetHandlers = () => {
  server.resetHandlers();
};

/**
 * Utility function to delay API responses (useful for testing loading states)
 * @param path API path to override
 * @param method HTTP method
 * @param responseData Data to return in the response
 * @param delayMs Delay in milliseconds
 * @param status HTTP status code
 */
export const mockDelayedResponse = (
  path: string,
  method: 'get' | 'post' | 'put' | 'delete' | 'patch',
  responseData: any,
  delayMs = 1000,
  status = 200
) => {
  const handler = rest[method](
    path,
    async (req, res, ctx) => {
      await new Promise(resolve => setTimeout(resolve, delayMs));
      return res(
        ctx.status(status),
        ctx.json(responseData)
      );
    }
  );

  server.use(handler);
};
