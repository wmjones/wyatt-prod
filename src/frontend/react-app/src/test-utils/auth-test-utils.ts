/**
 * Authentication testing utilities for React components
 */
import React from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { MemoryRouter, MemoryRouterProps } from 'react-router-dom';
// Import auth module
import * as authModule from '../utils/auth';

// Set up the mock implementation for Jest
jest.mock('../utils/auth', () => ({
  getCurrentUser: jest.fn(),
  signIn: jest.fn(),
  signUp: jest.fn(),
  confirmSignUp: jest.fn(),
  signOut: jest.fn(),
  getEnvConfig: jest.fn(),
  AuthState: {
    SIGNED_IN: 'SIGNED_IN',
    SIGNED_OUT: 'SIGNED_OUT',
    LOADING: 'LOADING',
    CONFIG_ERROR: 'CONFIG_ERROR'
  }
}));

// Import AuthState type from auth module for use in mocks
import { AuthState } from '../utils/auth';

/**
 * Mock user object type
 */
export interface MockUser {
  username: string;
  attributes: {
    email: string;
    sub: string;
    [key: string]: any;
  };
}

/**
 * Creates a mock user object
 */
export function createMockUser(overrides: Partial<MockUser> = {}): MockUser {
  return {
    username: 'testuser',
    attributes: {
      email: 'test@example.com',
      sub: 'test-user-id',
      ...(overrides.attributes || {}),
    },
    ...overrides,
  };
}

/**
 * Mock authentication options for testing
 */
export interface MockAuthOptions {
  isAuthenticated?: boolean;
  user?: MockUser;
  mockSignIn?: 'success' | 'error' | 'configError';
  mockSignUp?: 'success' | 'error' | 'configError';
  mockConfirmSignUp?: 'success' | 'error';
}

/**
 * Sets up authentication mocks for testing
 */
export function setupAuthMocks(options: MockAuthOptions = {}): void {
  const {
    isAuthenticated = false,
    user = createMockUser(),
    mockSignIn = 'success',
    mockSignUp = 'success',
    mockConfirmSignUp = 'success',
  } = options;

  // Mock getCurrentUser
  (authModule.getCurrentUser as jest.Mock).mockResolvedValue(isAuthenticated ? user : null);

  // Mock signIn based on option
  if (mockSignIn === 'success') {
    (authModule.signIn as jest.Mock).mockResolvedValue(user);
  } else if (mockSignIn === 'error') {
    (authModule.signIn as jest.Mock).mockRejectedValue(new Error('Invalid email or password'));
  } else if (mockSignIn === 'configError') {
    (authModule.signIn as jest.Mock).mockRejectedValue(
      new Error('Authentication service configuration error. Please contact support with error code: USER_POOL_CLIENT_ERROR')
    );
  }

  // Mock signUp based on option
  if (mockSignUp === 'success') {
    (authModule.signUp as jest.Mock).mockResolvedValue({ isSignUpComplete: false, nextStep: 'CONFIRM_SIGN_UP' });
  } else if (mockSignUp === 'error') {
    (authModule.signUp as jest.Mock).mockRejectedValue(new Error('This email is already registered. Please sign in instead.'));
  } else if (mockSignUp === 'configError') {
    (authModule.signUp as jest.Mock).mockRejectedValue(
      new Error('Authentication service configuration error. Please contact support with error code: USER_POOL_CLIENT_ERROR')
    );
  }

  // Mock confirmSignUp based on option
  if (mockConfirmSignUp === 'success') {
    (authModule.confirmSignUp as jest.Mock).mockResolvedValue(true);
  } else if (mockConfirmSignUp === 'error') {
    (authModule.confirmSignUp as jest.Mock).mockRejectedValue(new Error('Invalid verification code. Please try again.'));
  }

  // Always mock signOut to resolve successfully
  (authModule.signOut as jest.Mock).mockResolvedValue(undefined);

  // Mock getEnvConfig to return valid test values
  (authModule.getEnvConfig as jest.Mock).mockReturnValue({
    Auth: {
      Cognito: {
        userPoolId: 'mock-user-pool-id',
        userPoolClientId: 'mock-client-id'
      }
    }
  });
}

/**
 * Props for rendering with router and auth context
 */
export interface RenderWithAuthOptions extends RenderOptions {
  initialEntries?: MemoryRouterProps['initialEntries'];
  initialIndex?: MemoryRouterProps['initialIndex'];
  authOptions?: MockAuthOptions;
}

/**
 * Renders a component with router and auth context
 */
export function renderWithAuth(ui: React.ReactElement, options: RenderWithAuthOptions = {}) {
  const {
    initialEntries = ['/'],
    initialIndex = 0,
    authOptions = {},
    ...renderOptions
  } = options;

  // Setup auth mocks based on provided options
  setupAuthMocks(authOptions);

  // Helper to wrap component with all necessary providers
  const Wrapper = ({ children }: { children: React.ReactNode }) => {
    // Use React.createElement instead of JSX to avoid TypeScript errors in .ts files
    return React.createElement(
      MemoryRouter,
      { initialEntries, initialIndex },
      children
    );
  };

  // Return the render result with the provided wrapper
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}
