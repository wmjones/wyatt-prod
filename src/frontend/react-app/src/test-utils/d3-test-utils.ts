import { fireEvent, RenderResult, screen, waitFor } from '@testing-library/react';
import React from 'react';
// Import d3 with type checking disabled in tests
const d3 = require('d3');

/**
 * D3 Testing Utilities
 * These functions help with testing D3 visualizations in React components
 */

/**
 * Creates a mock ref suitable for d3 visualization testing
 * @returns A mocked React.RefObject with a real SVG element
 */
export function createMockD3Ref(): React.RefObject<SVGSVGElement> {
  const mockSvg = document.createElementNS('http://www.w3.org/2000/svg', 'svg');
  // Add some basic dimensions to the SVG
  mockSvg.setAttribute('width', '500');
  mockSvg.setAttribute('height', '300');

  // Add g elements that d3 commonly works with
  const gAxis = document.createElementNS('http://www.w3.org/2000/svg', 'g');
  gAxis.setAttribute('class', 'axis');
  mockSvg.appendChild(gAxis);

  const gContent = document.createElementNS('http://www.w3.org/2000/svg', 'g');
  gContent.setAttribute('class', 'content');
  mockSvg.appendChild(gContent);

  return { current: mockSvg };
}

/**
 * Sets up React.useRef mock for D3 testing
 * Call this in your test setup (beforeEach) to mock useRef
 * @returns The mock ref object that was created
 */
export function setupD3RefMock(): React.RefObject<SVGSVGElement> {
  const mockRef = createMockD3Ref();
  jest.spyOn(React, 'useRef').mockReturnValue(mockRef);
  return mockRef;
}

/**
 * Finds and returns a D3 selection within a component
 * @param selector CSS selector for the D3 element
 * @param container Container element (optional)
 * @returns D3 selection
 */
export function getD3Selection(
  selector: string,
  container: Element | null = document.body
): d3.Selection<d3.BaseType, unknown, null, undefined> {
  return d3.select(container).select(selector);
}

/**
 * Gets all D3 selections matching a selector
 * @param selector CSS selector for the D3 elements
 * @param container Container element (optional)
 * @returns D3 selection of multiple elements
 */
export function getAllD3Selections(
  selector: string,
  container: Element | null = document.body
): d3.Selection<d3.BaseType, unknown, null, undefined> {
  return d3.select(container).selectAll(selector);
}

/**
 * Gets a path's data from a D3 path element
 * @param pathSelector CSS selector for the path
 * @param container Container element (optional)
 * @returns Array of data points or null if path not found
 */
export function getD3PathData(
  pathSelector: string,
  container: Element | null = document.body
): { x: number, y: number }[] | null {
  const path = getD3Selection(pathSelector, container).node();
  if (!path) return null;

  // For a path with a datum (common in D3)
  return d3.select(path).datum() as { x: number, y: number }[] | null;
}

/**
 * Simulates a mouse hover on a D3 element
 * @param selector CSS selector for the D3 element to hover
 * @param x X coordinate for the hover
 * @param y Y coordinate for the hover
 * @param container Container element (optional)
 */
export function simulateD3MouseHover(
  selector: string,
  x: number,
  y: number,
  container: Element | null = document.body
): void {
  const element = container?.querySelector(selector) as SVGElement;
  if (!element) {
    throw new Error(`Element with selector "${selector}" not found`);
  }

  // Create a custom mouse event
  const mouseOverEvent = new MouseEvent('mouseover', {
    bubbles: true,
    cancelable: true,
    clientX: x,
    clientY: y
  });

  const mouseMoveEvent = new MouseEvent('mousemove', {
    bubbles: true,
    cancelable: true,
    clientX: x,
    clientY: y
  });

  // Dispatch the events
  element.dispatchEvent(mouseOverEvent);
  element.dispatchEvent(mouseMoveEvent);
}

/**
 * Simulates a mouse leave on a D3 element
 * @param selector CSS selector for the D3 element
 * @param container Container element (optional)
 */
export function simulateD3MouseLeave(
  selector: string,
  container: Element | null = document.body
): void {
  const element = container?.querySelector(selector) as SVGElement;
  if (!element) {
    throw new Error(`Element with selector "${selector}" not found`);
  }

  // Create a custom mouse event
  const mouseLeaveEvent = new MouseEvent('mouseleave', {
    bubbles: true,
    cancelable: true
  });

  // Dispatch the event
  element.dispatchEvent(mouseLeaveEvent);
}

/**
 * Gets attributes from a D3 element
 * @param selector CSS selector for the D3 element
 * @param attributes Array of attribute names to extract
 * @param container Container element (optional)
 * @returns Object with attribute name-value pairs
 */
export function getD3Attributes(
  selector: string,
  attributes: string[],
  container: Element | null = document.body
): Record<string, string | null> {
  const selection = getD3Selection(selector, container);
  const element = selection.node();

  if (!element) {
    throw new Error(`Element with selector "${selector}" not found`);
  }

  const result: Record<string, string | null> = {};
  attributes.forEach(attr => {
    result[attr] = (element as SVGElement).getAttribute(attr);
  });

  return result;
}

/**
 * Waits for a D3 transition to complete
 * @param container Container element with the D3 animation
 * @returns Promise that resolves when transitions are complete
 */
export async function waitForD3Transitions(
  container: Element | null = document.body
): Promise<void> {
  // Get all transitioning elements
  const transitioningElements = d3.select(container).selectAll('*').filter(function() {
    return d3.active(this);
  });

  // If no transitions, resolve immediately
  if (transitioningElements.empty()) {
    return Promise.resolve();
  }

  // Otherwise wait for transitions to complete
  // This is a simplified approach; actual transition time may vary
  return new Promise(resolve => {
    setTimeout(resolve, 1100); // Add a small buffer to the default 1000ms
  });
}

/**
 * Verifies the presence of specific D3 visualization components
 * @param container Container element
 * @param selectors Object mapping meaningful names to CSS selectors
 * @returns Object with boolean values indicating presence of each component
 */
export function checkD3Components(
  container: Element | null = document.body,
  selectors: Record<string, string>
): Record<string, boolean> {
  const results: Record<string, boolean> = {};

  Object.entries(selectors).forEach(([name, selector]) => {
    const element = container?.querySelector(selector);
    results[name] = !!element;
  });

  return results;
}

/**
 * Creates a basic D3 mock implementation for tests
 * Used to replace the d3 module in tests
 */
export function createD3Mock() {
  // Create a chainable mock function that returns itself
  const createChainableMock = () => {
    const mock = jest.fn();
    mock.mockReturnThis = function() {
      return this;
    };

    // Add methods that return the mock itself for chaining
    const methods = [
      'select', 'selectAll', 'append', 'attr', 'style', 'text', 'data', 'datum',
      'enter', 'exit', 'remove', 'call', 'on', 'transition', 'duration', 'ease',
      'filter', 'ticks', 'tickFormat', 'innerRadius', 'outerRadius', 'x', 'y',
      'y0', 'y1', 'domain', 'range', 'nice', 'curve'
    ];

    methods.forEach(method => {
      mock[method] = jest.fn().mockReturnValue(mock);
    });

    // Add special methods with custom returns
    mock.node = jest.fn().mockReturnValue({
      getBoundingClientRect: () => ({ width: 100, height: 100 }),
      getTotalLength: () => 100
    });

    mock.empty = jest.fn().mockReturnValue(false);

    return mock;
  };

  // Create the base chainable mock
  const chainableMock = createChainableMock();

  // Create mock d3 implementation
  return {
    select: jest.fn(() => chainableMock),
    selectAll: jest.fn(() => chainableMock),

    // Scale mocks
    scaleLinear: jest.fn(() => {
      const scale = jest.fn(value => value);
      scale.domain = jest.fn().mockReturnValue(scale);
      scale.range = jest.fn().mockReturnValue(scale);
      scale.nice = jest.fn().mockReturnValue(scale);
      return scale;
    }),

    // Random data functions
    randomNormal: jest.fn(() => jest.fn(() => 0.5)),

    // Transition management
    active: jest.fn(() => null),

    // Array functions
    min: jest.fn(arr => arr ? Math.min(...arr) : 0),
    max: jest.fn(arr => arr ? Math.max(...arr) : 1),
    extent: jest.fn(arr => arr ? [Math.min(...arr), Math.max(...arr)] : [0, 1]),

    // Shape generators
    line: jest.fn(() => {
      const lineGen = jest.fn(() => 'M0,0L1,1');
      lineGen.x = jest.fn().mockReturnValue(lineGen);
      lineGen.y = jest.fn().mockReturnValue(lineGen);
      lineGen.curve = jest.fn().mockReturnValue(lineGen);
      return lineGen;
    }),
    area: jest.fn(() => {
      const areaGen = jest.fn(() => 'M0,0L1,1L1,0Z');
      areaGen.x = jest.fn().mockReturnValue(areaGen);
      areaGen.y0 = jest.fn().mockReturnValue(areaGen);
      areaGen.y1 = jest.fn().mockReturnValue(areaGen);
      areaGen.curve = jest.fn().mockReturnValue(areaGen);
      return areaGen;
    }),

    // Axis generators
    axisBottom: jest.fn(() => createChainableMock()),
    axisLeft: jest.fn(() => createChainableMock()),

    // Curve types
    curveBasis: {},
    easeLinear: {},

    // Math functions
    range: jest.fn(n => Array.from({ length: n }, (_, i) => i)),

    // For pie charts
    pie: jest.fn(() => ({
      value: jest.fn().mockReturnThis(),
      sort: jest.fn().mockReturnThis(),
      startAngle: jest.fn().mockReturnThis(),
      endAngle: jest.fn().mockReturnThis(),
      padAngle: jest.fn().mockReturnThis()
    })),
    arc: jest.fn(() => ({
      innerRadius: jest.fn().mockReturnThis(),
      outerRadius: jest.fn().mockReturnThis(),
      startAngle: jest.fn().mockReturnThis(),
      endAngle: jest.fn().mockReturnThis(),
      padAngle: jest.fn().mockReturnThis()
    }))
  };
}
