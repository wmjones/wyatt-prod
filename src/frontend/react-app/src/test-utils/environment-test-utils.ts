/**
 * Test utilities for mocking environment-related functions
 */
import * as environmentModule from '../utils/environment';

/**
 * Mock the environment as development
 */
export const mockDevelopmentEnvironment = (): jest.SpyInstance => {
  return jest.spyOn(environmentModule, 'isDevelopment').mockReturnValue(true);
};

/**
 * Mock the environment as production
 */
export const mockProductionEnvironment = (): jest.SpyInstance => {
  return jest.spyOn(environmentModule, 'isDevelopment').mockReturnValue(false);
};

/**
 * Mock specific environment function with custom value
 * @param functionName The environment function to mock
 * @param value The return value for the function
 */
export const mockEnvironmentFunction = (
  functionName: keyof typeof environmentModule,
  value: boolean | string
): jest.SpyInstance => {
  return jest.spyOn(environmentModule, functionName).mockImplementation(() => value);
};

/**
 * Reset all environment mocks
 */
export const resetEnvironmentMocks = (): void => {
  jest.restoreAllMocks();
};
