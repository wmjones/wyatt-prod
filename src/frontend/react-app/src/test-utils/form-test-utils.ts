import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { UserEvent } from '@testing-library/user-event/dist/types/setup/setup';

/**
 * Type definition for form field data
 */
interface FormFields {
  [key: string]: string;
}

/**
 * Utility to fill all fields in a form
 *
 * @param user UserEvent instance
 * @param formFields Object with field names as keys and values to enter
 */
export const fillForm = async (
  user: UserEvent,
  formFields: FormFields
): Promise<void> => {
  for (const [fieldName, value] of Object.entries(formFields)) {
    // Try to find the input by name, placeholder, label text, or role
    const input = screen.getByRole('textbox', { name: new RegExp(fieldName, 'i') }) ||
                  screen.getByPlaceholderText(new RegExp(fieldName, 'i')) ||
                  screen.getByLabelText(new RegExp(fieldName, 'i')) ||
                  screen.getByTestId(fieldName);

    await user.clear(input);
    await user.type(input, value);
  }
};

/**
 * Utility to submit a form after filling it
 *
 * @param user UserEvent instance
 * @param formFields Object with field names as keys and values to enter
 * @param submitButtonText Text on the submit button (default: 'Submit')
 */
export const fillAndSubmitForm = async (
  user: UserEvent,
  formFields: FormFields,
  submitButtonText = 'Submit'
): Promise<void> => {
  // Fill out the form
  await fillForm(user, formFields);

  // Find and click the submit button
  const submitButton = screen.getByRole('button', { name: new RegExp(submitButtonText, 'i') });
  await user.click(submitButton);
};

/**
 * Utility to check form validation errors
 *
 * @param fieldName Name of the field to validate
 * @returns Object with functions to check validation state
 */
export const formValidation = (fieldName: string) => {
  return {
    /**
     * Check if field has an error
     */
    hasError: (): boolean => {
      try {
        // Look for error messages associated with this field
        const errorElement = screen.queryByText(new RegExp(`(${fieldName}|this field).*error`, 'i'));
        if (errorElement) return true;

        // Look for ARIA invalid state
        const field = screen.queryByLabelText(new RegExp(fieldName, 'i')) ||
                     screen.queryByPlaceholderText(new RegExp(fieldName, 'i')) ||
                     screen.queryByTestId(new RegExp(`${fieldName}`, 'i'));

        return field ? field.getAttribute('aria-invalid') === 'true' : false;
      } catch (e) {
        return false;
      }
    },

    /**
     * Get error message for field
     */
    getErrorMessage: (): string | null => {
      try {
        // First try finding by aria-describedby
        const field = screen.queryByLabelText(new RegExp(fieldName, 'i')) ||
                     screen.queryByPlaceholderText(new RegExp(fieldName, 'i')) ||
                     screen.queryByTestId(new RegExp(`${fieldName}`, 'i'));

        if (field) {
          const describedBy = field.getAttribute('aria-describedby');
          if (describedBy) {
            const errorElement = document.getElementById(describedBy);
            if (errorElement) return errorElement.textContent;
          }
        }

        // If not found, look for error text near the field
        const errorElement = screen.queryByText(new RegExp(`(${fieldName}|this field).*error`, 'i'));
        return errorElement ? errorElement.textContent : null;
      } catch (e) {
        return null;
      }
    }
  };
};

/**
 * Utility to select an option from a dropdown
 *
 * @param user UserEvent instance
 * @param selectName The name or label of the select element
 * @param optionText The text of the option to select
 */
export const selectOption = async (
  user: UserEvent,
  selectName: string,
  optionText: string
): Promise<void> => {
  // Find the select element
  const select = screen.getByRole('combobox', { name: new RegExp(selectName, 'i') }) ||
                screen.getByLabelText(new RegExp(selectName, 'i')) ||
                screen.getByTestId(new RegExp(`${selectName}`, 'i'));

  // Click to open the dropdown
  await user.click(select);

  // Find and click the option
  const option = screen.getByRole('option', { name: new RegExp(optionText, 'i') });
  await user.click(option);
};
