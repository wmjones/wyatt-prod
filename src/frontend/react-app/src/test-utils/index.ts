// Re-export all testing utilities
export * from './testing-library-utils';
export * from './api-test-utils';
export * from './d3-test-utils';
export * from './render-utils';
export * from './form-test-utils';

// Also re-export everything from testing-library to provide a single import source
export * from '@testing-library/react';
export * from '@testing-library/user-event';
export * from '@testing-library/jest-dom';
