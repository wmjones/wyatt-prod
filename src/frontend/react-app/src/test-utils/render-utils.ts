import { RenderResult } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { UserEvent } from '@testing-library/user-event/dist/types/setup/setup';

/**
 * Type combining render result with user-event for easier testing
 */
export interface RenderWithUserResult extends RenderResult {
  user: UserEvent;
}

/**
 * Utility function to wait for a specific amount of time
 * Sometimes needed for components with timeouts or debounce
 *
 * @param ms Time to wait in milliseconds
 * @returns Promise that resolves after specified time
 */
export const waitMs = (ms: number): Promise<void> => {
  return new Promise(resolve => setTimeout(resolve, ms));
};

/**
 * Utility to generate test IDs with consistent format
 *
 * @param component Component name
 * @param element Element name
 * @returns Formatted test ID string
 */
export const testId = (component: string, element: string): string => {
  return `${component}-${element}`.toLowerCase();
};

/**
 * Helper to check if an element is visible in the viewport
 *
 * @param element The DOM element to check
 * @returns Boolean indicating if element is visible
 */
export const isElementInViewport = (element: Element): boolean => {
  const rect = element.getBoundingClientRect();

  return (
    rect.top >= 0 &&
    rect.left >= 0 &&
    rect.bottom <= (window.innerHeight || document.documentElement.clientHeight) &&
    rect.right <= (window.innerWidth || document.documentElement.clientWidth)
  );
};

/**
 * Helper to create a mock response object
 *
 * @param data The data to include in the response
 * @param status HTTP status code
 * @returns Mocked response object
 */
export const createMockResponse = <T>(data: T, status = 200) => {
  return {
    data,
    status,
    ok: status >= 200 && status < 300,
    statusText: status === 200 ? 'OK' : `Error ${status}`,
    headers: new Headers({ 'Content-Type': 'application/json' }),
  };
};

/**
 * Helper to simulate form field validation errors
 *
 * @param field The field to check
 * @param message Error message
 */
export const simulateFieldValidationError = (
  field: HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement,
  message: string
): void => {
  field.setCustomValidity(message);
  field.reportValidity();
};

/**
 * Helper to clear field validation errors
 *
 * @param field The field to clear validation for
 */
export const clearFieldValidationError = (
  field: HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement
): void => {
  field.setCustomValidity('');
};
