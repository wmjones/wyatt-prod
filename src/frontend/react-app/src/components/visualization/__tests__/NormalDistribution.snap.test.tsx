import React from 'react';
import renderer from 'react-test-renderer';
import { createD3Mock } from '../../../test-utils/d3-test-utils';

// Import component after mocking
import NormalDistribution from '../NormalDistribution';

// Mock D3 using our utility function
// Use direct mock instead of imported function
jest.mock('d3', () => ({
  select: jest.fn().mockReturnThis(),
  selectAll: jest.fn().mockReturnThis(),
  append: jest.fn().mockReturnThis(),
  attr: jest.fn().mockReturnThis(),
  style: jest.fn().mockReturnThis(),
  text: jest.fn().mockReturnThis(),
  data: jest.fn().mockReturnThis(),
  enter: jest.fn().mockReturnThis(),
  exit: jest.fn().mockReturnThis(),
  remove: jest.fn().mockReturnThis(),
  on: jest.fn().mockReturnThis(),
  domain: jest.fn().mockReturnThis(),
  range: jest.fn().mockReturnThis(),
  scaleLinear: jest.fn().mockReturnValue({
    domain: jest.fn().mockReturnThis(),
    range: jest.fn().mockReturnThis(),
    nice: jest.fn().mockReturnThis()
  }),
  axisBottom: jest.fn().mockReturnThis(),
  axisLeft: jest.fn().mockReturnThis(),
  line: jest.fn().mockReturnValue({
    x: jest.fn().mockReturnThis(),
    y: jest.fn().mockReturnThis(),
    curve: jest.fn().mockReturnThis()
  }),
  randomNormal: jest.fn().mockReturnValue(() => 0),
  curveBasis: {},
}));

/**
 * Minimal, focused snapshot tests for D3 visualizations
 *
 * These snapshots are kept minimal and focus only on the structure of the
 * component, not the full rendering details of the D3 visualization.
 * We're testing:
 * 1. The basic structure of the component
 * 2. That changing props updates the relevant user-visible information
 */
describe('NormalDistribution Component Snapshots', () => {
  // Mock React.useRef to avoid errors in snapshot tests
  beforeEach(() => {
    jest.spyOn(React, 'useRef').mockReturnValue({
      current: document.createElementNS('http://www.w3.org/2000/svg', 'svg')
    });
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('renders with default props', () => {
    const tree = renderer
      .create(<NormalDistribution />)
      .toJSON();

    // Use a more consistent serialization approach by directly using the snapshot
    // Instead of JSON.parse(JSON.stringify()) which can cause format differences
    expect(tree).toMatchSnapshot();
  });

  it('renders with custom attribution', () => {
    const tree = renderer
      .create(<NormalDistribution updatedBy="Custom User" />)
      .toJSON();

    // Use a more consistent serialization approach by directly using the snapshot
    expect(tree).toMatchSnapshot();
  });

  // We only need 2 snapshot tests:
  // 1. Default rendering - the basic component structure
  // 2. Customized rendering - changing the user-visible elements
  // This avoids brittle, excessive snapshot testing
});
