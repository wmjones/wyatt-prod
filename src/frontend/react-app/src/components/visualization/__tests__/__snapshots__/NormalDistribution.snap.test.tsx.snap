// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NormalDistribution Component Snapshots renders with custom attribution 1`] = `
<div
  style={
    {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "width": "100%",
    }
  }
>
  <div
    style={
      {
        "maxWidth": "600px",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <svg
      className="chart"
      style={
        {
          "aspectRatio": "600/300",
          "height": "auto",
          "maxWidth": "600px",
          "width": "100%",
        }
      }
    />
  </div>
  <p
    style={
      {
        "color": "#666",
        "fontSize": "0.9em",
        "fontStyle": "italic",
        "marginTop": "5px",
      }
    }
  >
    Last updated by: 
    Custom User
  </p>
</div>
`;

exports[`NormalDistribution Component Snapshots renders with default props 1`] = `
<div
  style={
    {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "width": "100%",
    }
  }
>
  <div
    style={
      {
        "maxWidth": "600px",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <svg
      className="chart"
      style={
        {
          "aspectRatio": "600/300",
          "height": "auto",
          "maxWidth": "600px",
          "width": "100%",
        }
      }
    />
  </div>
  <p
    style={
      {
        "color": "#666",
        "fontSize": "0.9em",
        "fontStyle": "italic",
        "marginTop": "5px",
      }
    }
  >
    Last updated by: 
    System
  </p>
</div>
`;
